Auto-generate dirac points in PorousFlowPolyLineSink

rwcarlsen opened the pull request:
Fixes #15547.

rwcarlsen committed:
fix error handling+msg in recent mods to porous flow line sink

Fixes #15528. Ref #15528

rwcarlsen committed:
modify porous flow line sink to auto-fill points between line base and endpoint

rwcarlsen committed:
add line sink tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/566946/) on e16d8a3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15615/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15615/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 961cb3bc3ace3e7e150e4eafda83c55425517835`

WilkAndy commented:
I can certainly review this, but maybe not today - we'll see.   I think you need to include some markdown documentation - probably a simple paragraph added to the existing doco.

rwcarlsen committed:
update markdown docs

moosebuild commented:
Job [Create Linux Stack](https://civet.inl.gov/job/567263/) on 8925b8b : invalidated by @milljm

very odd behavior for initailizing a Singularity container... try again. Error: cat: /home/moosetest/singularity/bind_paths/release_binds/CentOS-7.6/PR-_CentOS-7.6/moose/build: No such file or directory

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/567259/) on 8925b8b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15615/site)

This comment will be updated on new commits.

rwcarlsen committed:
make sure dirac points writer outputs in consistent order

loganharbour commented on modules/porous_flow/src/dirackernels/PorousFlowLineGeometry.C:
```diff
@@ -129,6 +135,64 @@ PorousFlowLineGeometry::PorousFlowLineGeometry(const InputParameters & parameter
+
+  // add point for each cell the line passes through
+  Point p0(_line_base[1], _line_base[2], _line_base[3]);
+  Point p1 = p0 + _line_length * _line_direction / _line_direction.norm();
```
Add an extra check to see if this point exists in the mesh up front so that you get a more meaningful out of bounds error?

rwcarlsen replied:
Not a bad idea.

rwcarlsen committed:
ensure line_base is inside the mesh

rwcarlsen commented:
yo yo - this is all green and ready to go.

loganharbour commented on modules/porous_flow/test/src/vectorpostprocessors/DiracPointsWriter.C:
```diff
@@ -0,0 +1,44 @@
+  _xs.clear();
+  _ys.clear();
+  _zs.clear();
+  std::set<Point> points;
```
`std::set<Point>` for sorting?

rwcarlsen replied:
Yes - the sorting matters for testing purposes - otherwise depending on the platform, you could get the points outputted in an inconsistent order - which actually caused test failures initially.

rwcarlsen commented:
@WilkAndy - any comments here before we merge this?

WilkAndy commented:
If you need this merged ASAP, go ahead.  I will only have time to review this in about 27 hours from now.

rwcarlsen commented:
It's not urgently necessary - for our immediate need, we are just using patched moose internally.  I can wait a day.

rwcarlsen commented:
It's not urgently necessary - for our immediate need, we are just using patched moose internally.  I can wait a day.

WilkAndy reviewed (dismissed):
I wish i could have done this when i made the Dirac Kernels initially, thanks @rwcarlsen .    Without actually testing it myself, i think it looks good.   Two questions:

- have you changed any of the existing test files to see whether they still pass with this new scheme.  that'd be great to see.

- what happens if the line goes out of the mesh?   i'm pretty sure that with the old method MOOSE used to crash with a pretty uninformative error.

rwcarlsen commented:
One of the tests was created by copying an existing test and gold file.  But it was a one-point-file case.  I haven't tried this with one of the multi-point files.

The problem with checking for the line going outside the mesh is that you already have use cases where you specify a line length in the third dimension of 2D problems - i.e. a simulated line source that really is like a point source.  If we add an error check for line end points out of the mesh then these cases won't work.

WilkAndy commented:
> One of the tests was created by copying an existing test and gold file. But it was a one-point-file case. I haven't tried this with one of the multi-point files.
> 
> The problem with checking for the line going outside the mesh is that you already have use cases where you specify a line length in the third dimension of 2D problems - i.e. a simulated line source that really is like a point source. If we add an error check for line end points out of the mesh then these cases won't work.

Ha, i never thought about that 2D-3D thing and the line going out of the mesh.    Well, it doesn't matter a lot, providing the error message is understandable in the case the line does exit the mesh.

rwcarlsen commented:
@WilkAndy - Are you saying to make the message an error?  We could make it a warning at least I suppose.  If we make it an error, than the test where I just used the same gold file from an existing test wouldn't be possible to run - or at least it would have to be rewritten - which sort of defeats the purpose of confirming that the auto-line code paths generate the same results as the manual points file.

WilkAndy commented:
Some sort of message is necessary.  i would actually make it an error, just in case the user didn't notice the warning message, but i'll leave that up to you.

rwcarlsen committed:
add line-exits-mesh error message and test

rwcarlsen commented:
Okay - I think I've worked out an okay set of errors/tests now.  I added the line-exits-mesh error requested and modified the test to not trigger it incorrectly.  This should be good to go now.

WilkAndy commented:
Great, thanks @rwcarlsen .   Let's merge this and see some people using it!

a

rwcarlsen commented:
@milljm - I carefully investigated the app failure - it is in sockeye where the max ad dofs per elem is being violated.  This is unrelated to my PR which only modifies the porous flow module.  The sockeye failures also pre-existed in the next branch and other PRs.
